README
NAME
    WWW::Splunk - Client library for Splunk log search engine

SYNOPSIS
      use WWW::Splunk;

      my $splunk = WWW::Splunk->new({
              host => $host,
              port => $port,
              login => $login,
              password => $password,
              unsafe_ssl => 1,
              verbose => 0,
      });

      my $sid = $splunk->start_search('selinux avc');
      $splunk->poll_search($sid);
      until ($splunk->results_read($sid)) {
              print scalar $splunk->search_results($sid);
      }
      print " results found\n";

DESCRIPTION
    This module contains utility functions for Splunk API, implementing
    version 4.1 API, verified to work with 4.2, 4.3 and 5.0.3 versions.

  start_search (string) [(since)] [(until)]
    Initiate a search, return a SID (Search ID) string.

  rt_search (string) (callback) [(since)] [(until)]
    Initiate a real-time search, calling a callback for each line matched.

    Finishes only if connection terminates (potentially never), returning
    number of results consumed.

  search_done (sid)
    Return true if the search is finished.

  poll_search (sid)
    Wait for a search to finish.

  search_results (sid)
    Return an array of the matched events. If called multiple times, it only
    returns events which were added from the time of the last call. Oh, and
    you can't run multiple search concurrently with single WWW::Splunk
    instance. Otherwise, WWW::Splunk is perfectly thread-safe.

  results_read (sid)
    Return true if search is finished and all there are no more results to
    read (everything was fetched with search_results).

AUTHORS
    Lubomir Rintel, <lkundrak@v3.sk>, Michal Josef Špaček <skim@cpan.org>

    The code is hosted on GitHub
    <http://github.com/tupinek/perl-WWW-Splunk>. Bug fixes and feature
    enhancements are always welcome.

LICENSE
     This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

DONATIONS
    If you use this library and find it useful, donations are greatly
    appreciated!

 Use https://liberapay.com/skim/donate


Changes
2.10

2.09 2023-12-11T09:50:11+01:00
 [CODE]
 - Add possibility of getting user/password from ENV variables.
 - Add possibility of usage of e.g. »|inputloookup __table__« Splunk query
   by adding backward compatible 'search' constructor variable which is
   adding 'search' keyword to each Splunk query.
 - Move hardcoded prefix=/services variable to 'service_prefix' constructor
   variable.
 - Remove spaces before subroutine/method call in bin/sc.
 - Retab in bin/sc.

 [DOC]
 - Add constructor default values to WWW::Splunk::API.
 - Add donation.

 [OTHER]
 - Add bugtracker.
 - Fix dependencies.
 - Fix test dependencies. We don't need dependencies for author tests.
 - Update author github username.

2.08 2017-11-07T10:07:05+01:00
 [FIXES]
 - Fix dependencies.

2.07 2017-11-06T17:09:52+01:00
 [DOC]
 - Rewrite to one style
 - Rewrite Changelog to my Changes

 [FIXES]
 - Fix bug with second search on object.
 - Remove dependency to obsolete Text::CSV.

 [TESTS]
 - Rewrite tests to recursive way, dir per module

2.06 2017-04-08T00:27:12+02:00
 [DOC]
 - Add LICENSE file
 - Add LICENSE header to doc
 - Add minimal version of Perl
 - Add README file
 - Fix email in doc

2.05 2017-04-07T12:17:03+02:00
 [CODE]
 - Add support for Kerberos (e.g. via LWP::Authen::Negotiate)
 - Add verbose parameter

 [DOC]
 - Change maintainer and repository

 [FIXES]
 - Fix documentation
 - Fix get calling
 - Proper handling of HTTP 204 response
 - Remove extra check for Client-SSL-Warning header

2.04 2013-08-20T17:29:57+02:00
 [FIXES]
 - Drop META.yml from the repository
 - Make parser return the raw XML document if it could not be parsed

2.03 2013-08-20T16:07:15+02:00
 [DOC]
 - 5.0.3 Splunk seems to work
 - Update changelog

 [FIXES]
 - Fix dealing with empty XML responses

2.02 2012-06-14T10:24:25+02:00
 [FIXES]
 - Do not attempt to interpret response after request is finished if it was
   already handled asynchronously
 - Sane defaults for host/port API usage
 - Fix documentation
 - Fix insecure magic

2.01 2012-05-10T11:45:31+02:00
 [FIXES]
 - Fix a warning
 - Fix raw value parsing

2.0 2012-05-09T00:29:47+02:00
 [CODE]
 - Add utility support for real-time searches
 - Add library support for real-time searches
 - Drop use of CSV API, replace with XML
 - XMLParser: Handle more type of XML documents
 - Add support for different output formats
 - Extra checking of a response

 [DOC]
 - Update versioning information.

 [FIXES]
 - Remove a debugging leftover
 - Kill a warning

1.11 2011-09-25T17:13:15+02:00
 [CODE]
 - Add support for limiting the search query with a time frame

 [DOC]
 - Nicer POD inlining

 [FIXES]
 - HTML responses do not contain response messages

1.10 2010-12-12T12:40:21+01:00
 [CODE]
 - WWW::Splunk: Read results in bigger batches to improve performance

 [FIXES]
 - WWW::Splunk: Read processed results instead of raw events so that
   aggregates work [rt#59231]
 - WWW::Splunk: Add results_read() method to reliably detect when results
   are read.

1.03 2010-05-05T08:21:18+02:00
 [CODE]
 - WWW::Splunk::XMLParser: Do not attempt to grok arbitrary XML from
   Splunk, return XML::LibXML instance instead
 - WWW::Splunk: Print error messages from splunk as well

 [FIXES]
 - sc: Print the result of the last search iteration as well

1.02 2010-04-28T18:09:52+01:00
 [CODE]
 - Rename Net::Splunk to WWW::Splunk

1.01 2010-04-27T21:04:00+01:00
 [FIXES]
 - Add a missing shebang to the launcher script.

1.00 2010-04-27T20:36:54+01:00
 - Initial release with search capability.


Build.PL
use strict;
use warnings;

use Module::Build;

my $build = Module::Build->new(
	module_name => 'WWW::Splunk',
	license => 'perl',
	dist_author => 'Michal Josef Spacek <skim@cpan.org>',
	dist_version => '2.10',
	requires => {
		'perl' => 5.010,
		'Carp' => 0,
		'Date::Manip' => 0,
		'Getopt::Long' => 0,
		'HTTP::Request::Common' => 0,
		'LWP::UserAgent' => 0,
		'Pod::Usage' => 0,
		'XML::LibXML' => 0,
	},
	build_requires => {
		'Test::More' => 0,
		'XML::LibXML' => 0,
	},
	test_requires => {
		'Test::More' => 0,
		'Test::NoWarnings' => 0,
		'XML::LibXML' => 0,
	},
	configure_requires => {
		'Module::Build' => 0.42,
	},
	meta_merge => {
		resources => {
			bugtracker => 'https://github.com/michal-josef-spacek/perl-WWW-Splunk/issues',
			repository => 'https://github.com/michal-josef-spacek/perl-WWW-Splunk',
		},
	},
	recursive_test_files => 1,
);

$build->create_build_script;


lib/WWW/Splunk/API.pm
=encoding utf8

=head1 NAME

WWW::Splunk::API - Splunk REST client

=head1 DESCRIPTION

L<WWW::Splunk::API> is a low-level interface to Splunk
log search engine. It deals with HTTP communication as well as
working around certain interface glitches.

See L<http://www.splunk.com/base/Documentation/latest/Developer/RESTSearch>
for API definition.

This module is designed to be Splunk API version agnostic.

=cut

package WWW::Splunk::API;

use LWP::UserAgent;
use HTTP::Request::Common;
use WWW::Splunk::XMLParser;
use Carp;

use strict;
use warnings;

our $VERSION = '2.10';

=head2 B<new> (F<params>)

A constructor.

  my $splunk = WWW::Splunk::API->new({
          host => $host,
          port => $port,
          login => $login,
          password => $password,
          unsafe_ssl => 0,
          verbose => 0,
  });

Default values are:
 - port - 8089
 - host - localhost
 - url - https://$host:$port
 - verbose - 0
 - unsafe_ssl - 0
 - service_prefix - /services
 - search - 1

=cut

sub new {
	my ($class, $self) = @_;

	$self->{port} ||= 8089;
	$self->{host} ||= 'localhost';
	$self->{url} ||= 'https://'.$self->{host}.':'.$self->{port};
	$self->{verbose} ||= 0;
	$self->{service_prefix} ||= '/services';
	$self->{search} ||= 1;

	# Set up user agent unless an existing one was passed
	unless ($self->{agent}) {
		$self->{agent} = LWP::UserAgent->new(
			ssl_opts =>  {verify_hostname => (not $self->{unsafe_ssl})},
		);
		$self->{agent}->cookie_jar ({});
		$self->{agent}->credentials(
			delete ($self->{host}).':'.(delete $self->{port}),
			'/splunk',
			delete $self->{login},
			delete $self->{password},
		) if exists $self->{login};
		$self->{agent}->agent("$class/$VERSION ");
	}

	return bless $self, $class;
}

=head2 B<delete> (F<parameters>)

Wrapper around HTTP::Request::Common::DELETE().

=cut

sub delete {
	my ($self, @args) = @_;

	print "DELETE" if $self->{verbose};
	$self->request(\&DELETE, @args);
}

=head2 B<post> (F<parameters>)

Wrapper around HTTP::Request::Common::POST().

=cut

sub post {
	my ($self, @args) = @_;

	print "POST" if $self->{verbose};
	$self->request(\&POST, @args);
}

=head2 B<get> (F<parameters>)

Wrapper around HTTP::Request::Common::GET().

=cut

sub get {
	my ($self, @args) = @_;

	print "GET" if $self->{verbose};
	$self->request(\&GET, @args);
}

=head2 B<head> (F<parameters>)

Wrapper around HTTP::Request::Common::HEAD().
Not used anywhere in splunk API

=cut

sub head {
	my ($self, @args) = @_;

	print "HEAD" if $self->{verbose};
	$self->request(\&HEAD, @args);
}

=head2 B<put> (F<parameters>)

Wrapper around HTTP::Request::Common::PUT().
Not used anywhere in splunk API

=cut

sub put {
	my ($self, @args) = @_;

	print "PUT" if $self->{verbose};
	$self->request(\&PUT, @args);
}

=head2 B<request> (F<method>, F<location>, [F<data>], [F<callback>])

Request a Splunk api and deal with the results.

Method can be either a L<HTTP::Request> instance (see L<HTTP::Request::Common>
for useful ones), or a plain string, such as "GET" or "DELETE."

Optional F<data> is has reference gets serialized into a request body for POST
request. Use I<undef> in case you don't have any data to send, but need to
specify a callback function in subsequent argument.

Call-back function can be specified for a single special case, where a XML stream
of <results> elements is expected.

=cut

sub request {
	my ($self, $method, $location, $data, $callback) = @_;

	my $url = $self->{url}.$self->{service_prefix}.$location;
	if ($self->{verbose}) {
		print " $url\n";
		if (defined $data) {
			foreach my $key (sort keys %$data) {
				my $value = $data->{$key};
				$value =~ s/\n/ /msg;
				print "- $key => $value\n";
			}
		}
	}

	# Construct the request
	my $request;
	if (ref $method and ref $method eq 'CODE') {
		# Most likely a HTTP::Request::Common
		if (! defined $data) {
			$request = $method->($url);
		} else {
			$request = $method->($url, $data);
		}
	} else {
		# A method string
		$request = HTTP::Request->($method, $url);
	}

	my $content_type = '';
	my $buffer;

	$self->{agent}->remove_handler('response_header');
	$self->{agent}->add_handler(response_header => sub {
		my ($response, $ua, $h) = @_;

		# Do not think of async processing of error responses
		return 0 unless $response->is_success;

		my $content_type_header = $response->header('Content-Type') // '';
		if ($content_type_header =~ /^([^\s;]+)/) {
			$content_type = $1;
		} elsif ($response->code ne 204) {
			# Sometimes splunk return HTTP 204 NO CONTENT during poll_search() call,
			# Content-Type header is empty in this case. We must not croak in this case.
			croak "Missing or invalid Content-Type: $content_type_header";
		}

		if ($callback) {
			$response->{default_add_content} = 0;
			$buffer = "";
		}
	});

	$self->{agent}->remove_handler('response_data');
	$self->{agent}->add_handler(response_data => sub {
		my ($response, $ua, $h, $data) = @_;

		return 1 unless defined $buffer;
		$buffer .= $data;
		foreach (split /<\/results>\K/, $buffer) {
			unless (/<\/results>$/) {
				$buffer = $_;
				last;
			}

			my $xml = XML::LibXML->load_xml(string => $_);
			$callback->(WWW::Splunk::XMLParser::parse($xml));
		}

		return 1;
	}) if $callback;

	# Run it
	my $response = $self->{agent}->request($request);
	croak $response->header ('X-Died') if $response->header ('X-Died');

	# Deal with HTTP errors
	unless ($response->is_success) {
		my $content = WWW::Splunk::XMLParser::parse ($response->content)
			if $response->header ('Content-Type') =~ /xml/;
		my $error = "HTTP Error: ".$response->status_line;
		$error .= sprintf "\n%s: %s",
			$content->findvalue ('/response/messages/msg/@type'),
			$content->findvalue ('/response/messages/msg')
			if eval { $content->isa ('XML::LibXML::Document') }
				and $content->documentElement->nodeName eq 'response';
		croak $error;
	}

	# We've gotten the response already
	return if $callback;

	# Parse content from synchronous responses
	# TODO: use callback and m_media_type matchspecs
	if ($content_type eq 'text/xml') {
		my $xml = XML::LibXML->load_xml (string => $response->content);
		my @ret = WWW::Splunk::XMLParser::parse ($xml);
		return $#ret ? @ret : $ret[0];
	} elsif ($response->code eq 204) {
		# "No content"
		# Happens when events are requested immediately
		# after the job is enqueued. With a text/plain content type
		# Empty array is the least disturbing thing to return here
		return ();
	} elsif ($content_type eq 'text/plain') {
		# Sometimes an empty text/plain body is sent
		# even without 204 return code.
		return ();
	} else {
		# TODO: We probably can't do much about RAW
		# format, yet we could parse at least JSON
		croak "Unknown content type: $content_type";
	}
}

=head1 SEE ALSO

L<WWW::Splunk>, L<sc>

=head1 AUTHORS

Lubomir Rintel, L<< <lkundrak@v3.sk> >>,
Michal Josef Špaček L<< <skim@cpan.org> >>

The code is hosted on GitHub L<http://github.com/michal-josef-spacek/perl-WWW-Splunk>.
Bug fixes and feature enhancements are always welcome.

=head1 LICENSE

 This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut


lib/WWW/Splunk/XMLParser.pm
=encoding utf8

=head1 NAME

WWW::Splunk::XMLParser - Parse Splunk XML format

=head1 DESCRIPTION

This is an utility module to deal with XML format ocassionally returned
by Splunk and seemlingly undocumented.

Note that Splunk usually returns Atom XMLs, which have the same
content type. They can be distinguished by a DOCTYPE.

=cut

package WWW::Splunk::XMLParser;

use strict;
use warnings;

use XML::LibXML qw/:libxml/;
use Carp;

our $VERSION = '2.10';

=head2 B<parse> (F<string>)

Return a perl structure from a XML string, if it's
parsable, otherwise return a raw XML::LibXML object

=cut

sub parse {
	my $xml = shift;

	my @tree = eval { parsetree ($xml) };
	return $xml if $@;
	return $#tree ? @tree : $tree[0];
}

=head2 B<parsetree> (F<XML::LibXML::Node>)

Parse a XML node tree recursively.

=cut

sub parsetree {
	my $xml = shift;
	my @retval;

	my $has_elements = grep { $_->nodeType eq XML_ELEMENT_NODE }
		$xml->nonBlankChildNodes ();

	foreach my $node ($xml->nonBlankChildNodes ()) {

		# Not interested in anything but elements
		next if $has_elements and $node->nodeType ne XML_ELEMENT_NODE;

		# Structure or structure wrapped in Atom
		if ($node->nodeName () eq 'list' or
			$node->nodeName () eq 's:list') {
			push @retval, [ parsetree ($node) ];
		} elsif ($node->nodeName () eq 'dict' or
			$node->nodeName () eq 's:dict') {
			push @retval, { parsetree ($node) };
		} elsif ($node->nodeName () eq 'key' or
			$node->nodeName () eq 's:key') {
			push @retval, $node->getAttribute ('name')
				=> scalar parsetree($node);
		} elsif ($node->nodeName () eq 'response' or
			$node->nodeName () eq 'item' or
			$node->nodeName () eq 's:item') {
			# Basically just ignore these
			push @retval, parsetree ($node);
		} elsif ($node->nodeName () eq 'entry') {
			# Crippled Atom envelope
			foreach my $node ($node->childNodes ()) {
				return parsetree ($node) if $node->nodeName () eq 'content';
			}
		} elsif ($node->nodeType eq XML_TEXT_NODE or $node->nodeName () eq '#cdata-section') {
			return $node->textContent;

		# Results
		} elsif ($node->nodeName () eq 'results') {
			return map { { parsetree ($_) } }
				grep { $_->nodeName eq 'result' }
				$node->childNodes;
		} elsif ($node->nodeName () eq 'field') {
			push @retval, $node->getAttribute ('k')
				=> scalar parsetree($node);
		} elsif ($node->nodeName () eq 'value'
			or $node->nodeName () eq 'v') {
			return $node->textContent;

		# Errors
		} else {
			die "Unknown XML element: ".$node->nodeName
		}
	}

	return wantarray ? @retval : $retval[0];
}

=head1 SEE ALSO

L<WWW::Splunk>, L<WWW::Splunk::API>, L<XML::LibXML>

=head1 AUTHORS

Lubomir Rintel, L<< <lkundrak@v3.sk> >>,
Michal Josef Špaček L<< <skim@cpan.org> >>

The code is hosted on GitHub L<http://github.com/michal-josef-spacek/perl-WWW-Splunk>.
Bug fixes and feature enhancements are always welcome.

=head1 LICENSE

 This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=cut

1;


lib/WWW/Splunk.pm
=encoding utf8

=head1 NAME

WWW::Splunk - Client library for Splunk log search engine

=head1 SYNOPSIS

  use WWW::Splunk;

  my $splunk = WWW::Splunk->new({
          host => $host,
          port => $port,
          login => $login,
          password => $password,
          unsafe_ssl => 1,
          verbose => 0,
  });

  my $sid = $splunk->start_search('selinux avc');
  $splunk->poll_search($sid);
  until ($splunk->results_read($sid)) {
          print scalar $splunk->search_results($sid);
  }
  print " results found\n";

=head1 DESCRIPTION

This module contains utility functions for Splunk API, implementing
version 4.1 API, verified to work with 4.2, 4.3 and 5.0.3 versions.

=cut

package WWW::Splunk;

use strict;
use warnings;

our $VERSION = '2.10';

use WWW::Splunk::API;
use Carp;
use Date::Manip;

use base qw/WWW::Splunk::API/;

=head2 B<start_search> (F<string>) [(F<since>)] [(F<until>)]

Initiate a search, return a SID (Search ID) string.

=cut

sub start_search {
	my ($self, $string, $since, $until) = @_;

	delete $self->{last_read};

	# Format dates
	($since, $until) = map { defined $_ ? scalar UnixDate (ParseDate ($_) || $_, '%O') || $_ : undef }
		($since, $until);

	if ($self->{search}) {
		$string = 'search '.$string;
	}

	$self->{results_consumed} = 0;
	my $response = $self->post ('/search/jobs', {
		search => "$string",
		(defined $since ? (earliest_time => $since) : ()),
		(defined $until ? (latest_time => $until) : ()),
	});
	die 'Unexpected response format '
		unless $response and ref $response eq 'XML::LibXML::Document';
	my $sid = $response->findvalue ('/response/sid');
	croak 'Bad response' unless defined $sid;

	return $sid;
}

=head2 B<rt_search> (F<string>) (F<callback>) [(F<since>)] [(F<until>)]

Initiate a real-time search, calling a callback for each line matched.

Finishes only if connection terminates (potentially never), returning number of
results consumed.

=cut

sub rt_search {
	my ($self, $string, $callback, $since, $until, $counter) = @_;

	$since ||= 'rt';
	$until ||= 'rt';
	$counter ||= 0;

	$self->post ('/search/jobs/export', {
		search => "search $string",
		earliest_time => $since,
		latest_time => $until,
		search_mode => 'realtime',
	}, sub {
		$callback->(@_);
		$counter++;
	});

	return $counter;
}

=head2 B<search_done> (F<sid>)

Return true if the search is finished.

=cut

sub search_done {
	my ($self, $sid) = @_;

	my $search = $self->get ('/search/jobs/'.$sid);

	return $search->{isDone};
}

=head2 B<poll_search> (F<sid>)

Wait for a search to finish.

=cut

sub poll_search {
	my ($self, $sid) = @_;

	until ($self->search_done ($sid)) {
		sleep 1;
	}
}

=head2 B<search_results> (F<sid>)

Return an array of the matched events.
If called multiple times, it only returns events which
were added from the time of the last call.
Oh, and you can't run multiple search concurrently
with single L<WWW::Splunk> instance. Otherwise,
L<WWW::Splunk> is perfectly thread-safe.

=cut

sub search_results {
	my ($self, $sid) = @_;

	my $done = $self->search_done ($sid);
	my @results = $self->get ('/search/jobs/'.$sid.'/results?count=1024&offset='.
		$self->{results_consumed});
	$self->{results_consumed} += scalar @results;
	$self->{last_read} = scalar @results if $done;

	return @results;
}

=head2 B<results_read> (F<sid>)

Return true if search is finished and all there are no
more results to read (everything was fetched with L<search_results>).

=cut

sub results_read {
	my ($self, $sid) = @_;

	return undef if not defined $self->{last_read};
	return $self->{last_read} eq 0;
}

=head1 AUTHORS

Lubomir Rintel, L<< <lkundrak@v3.sk> >>,
Michal Josef Špaček L<< <skim@cpan.org> >>

The code is hosted on GitHub L<http://github.com/michal-josef-spacek/perl-WWW-Splunk>.
Bug fixes and feature enhancements are always welcome.

=head1 LICENSE

 This is free software; you can redistribute it and/or modify it under the same terms as the Perl 5 programming language system itself.

=head1 DONATIONS

If you use this library and find it useful, donations are greatly appreciated!

=begin text

 Use https://liberapay.com/skim/donate

=end text

=begin html

<a href="https://liberapay.com/skim/donate">
  <img alt="Donate using Liberapay"
    src="https://liberapay.com/assets/widgets/donate.svg">
</a>

=end html

=cut

1;
